f3df/src/renderables.rs

```diff
@@ -0,0 +1,203 @@
+    center: &Vector3,
+    offset: &Vector3,
+    scale: &Vector3,
+    rotation_matrix: &Rotation3,
+) -> Matrix4 {
+    let translation_matrix = Translation3::from(*center);
+    let offset_matrix = Translation3::from(*offset);
+    let scale_matrix = Matrix4::new_nonuniform_scaling(scale);
+    // NOTE rotation purposefully happens first to make the face point in the correct direction
+    // TODO remove Matrix4::from if possible
+    Matrix4::from(translation_matrix) * scale_matrix * Matrix4::from(offset_matrix) * Matrix4::from(*rotation_matrix)
+}
+
+pub fn convert_sector(sector: &f3df::Sector) -> Vec<Face> {
+    let mut instance_data = Vec::new();
+    let contents = match &sector.sector_contents {
+        Some(x) => x,
+        None => return Vec::new(),
+    };
+    let origin = contents.grid_origin;
+    let cell_size = [
+        u64::from(contents.grid_size[0] - 1),
+        u64::from(contents.grid_size[1] - 1),
+        u64::from(contents.grid_size[2] - 1),
+    ];
+    let cell_size_i_j = cell_size[0] * cell_size[1];
+    let increment = contents.grid_increment;
+    let origin = Vector3::new(origin[0], origin[1], origin[2]);
+    for node in &contents.nodes {
+        let color = node.color;
+        let compress_type = node.compress_type;
+        for face in &node.faces {
+            let cell_index = face.index as u64;
+            let cell_index_i_j = cell_index % cell_size_i_j;
+            let i = cell_index_i_j % (cell_size[0]);
+            let j = cell_index_i_j / (cell_size[0]);
+            let k = cell_index / (cell_size_i_j);
+            let x = i as f32 + 0.5;
+            let y = j as f32 + 0.5;
+            let z = k as f32 + 0.5;
+
+            let center = Vector3::new(
+                origin[0] + increment * x,
+                origin[1] + increment * y,
+                origin[2] + increment * z,
+            );
+
+            let count = (u32::from(face.repetitions) + 1) as f32;
+            if face
+                .face_flags
+                .intersects(f3df::FaceFlags::POSITIVE_X_VISIBLE)
+            {
+                let offset = Vector3::new(1.0, 0.5, 0.5);
+                let normal = Vector3::new(1.0, 0.0, 0.0);
+                let rotation = Rotation3::from_axis_angle(&Vector3::y_axis(), FRAC_PI_2);
+                let scale = if compress_type.intersects(f3df::CompressFlags::POSITIVE_X_REPEAT_Y) {
+                    Vector3::new(increment, increment, count * increment)
+                } else {
+                    Vector3::new(increment, count * increment, increment)
+                };
+                instance_data.push(Face {
+                    color: [
+                        f32::from(color[0]) / 255.0,
+                        f32::from(color[1]) / 255.0,
+                        f32::from(color[2]) / 255.0,
+                    ],
+                    normal,
+                    matrix: compose_face_matrix(&center, &offset, &scale, &rotation),
+                });
+            }
```

Comment:
Can this be a separate function accepting offset, normal, rotationAxis, rotationAngle, FaceFlags. CompressFlags or something similar?